aesd-char-driver/main.c

```diff
@@ -0,0 +1,267 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ * @modifed_by Daniel Mendez for AESD Assignment 8
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include "aesdchar.h"
+
+#define TEMP_BUFFER_SIZE 0
+#define TEMP_CHUNK_SIZE 40
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Daniel Mendez"); 
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+char * temp_buffer = NULL;
+size_t temp_buffer_size = 0;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    PDEBUG("open");
+	//Setup fil-->private data
+	
+	struct aesd_dev *dev;
+	
+	dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+	filp->private_data = dev;
+	
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+	
+	//Nothing to do as the memory was allocated in init_module
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    struct aesd_dev *dev = (struct aesd_dev *)filp->private_data;
+    struct aesd_circular_buffer *circular_buffer = &dev->circular_buffer;
+    struct aesd_buffer_entry *entry = NULL;
+    size_t bytes_to_read;
+    ssize_t retval = 0;
+    size_t entry_offset_byte;
+    
+    if (buf == NULL) {
+        retval = -EFAULT;
+        goto exit;
+    }
+
+    mutex_lock(&dev->lock);
+
+
+    entry = aesd_circular_buffer_find_entry_offset_for_fpos(circular_buffer, *f_pos, &entry_offset_byte);
+    
+    if(entry == NULL) {
+        retval = 0;
+        goto exit;
+    }
+
+    bytes_to_read = entry->size - entry_offset_byte;
+
+    if (bytes_to_read > count) {
+        bytes_to_read = count;
+    }
+
+
+    if (copy_to_user(buf, entry->buffptr + entry_offset_byte, bytes_to_read)) {
+        retval = -EFAULT;
+        goto exit;
+    }
+
+    *f_pos += bytes_to_read;
+    retval = bytes_to_read;
+
+    exit:
+        mutex_unlock(&dev->lock);
+        return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+	int res;
+	int bytes_scanned = 0;
+    ssize_t retval = -ENOMEM;
+	
+	
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    
+	//Get the aesd device structure
+	struct aesd_dev *dev = filp->private_data;
+	//First check to see if there is a partial command ongoing
+	//Lock the mutex
+	res = mutex_lock_interruptible(&(dev->lock));
```

Comment:
@Silex93 you shouldn't need to lock until after line 127.  After that you need a lock for your global temp_buffer (but it would be better if that was inside the dev structure to make it clear what the lock was locking)